### PR TITLE
Refactored GradNodeAccumulation data structure and behaviour

### PR DESCRIPTION
**Pull request by jim19930609**
### PR types
New features

### PR changes
Others

### Describe
We used to register RetainGrad hook for leaf tensor so as to update grad tensor's value. After this patch, we asked GradNodeAccumulation to track a weak_ptr of grad_tensor, the value of which will get updated upon calling GradNodeAccumulation::operator(), instead of having to register a RetainGrad hook.

### CODE REVIEW

File: paddle/fluid/pybind/eager_method.cc [hunk 1]
```diff
@@ -199,17 +199,12 @@ static PyObject* tensor_clear_gradient(TensorObject* self, PyObject* args,
...
-    PADDLE_ENFORCE(
-        grad_node.get() != nullptr,
-        paddle::platform::errors::Fatal("Detected NULL grad_node"
-                                        "Leaf tensor should have had grad_node "
-                                        "with type: GradNodeAccumulation"));
-    auto accumulation_grad_node =
-        std::dynamic_pointer_cast<egr::GradNodeAccumulation>(grad_node);
-    grad = accumulation_grad_node->Grad();
+    grad = egr::EagerUtils::mutable_grad(self->tensor);
+    PADDLE_ENFORCE(grad != nullptr,
```
Comment (chenwhql) [inline paddle/fluid/pybind/eager_method.cc:203]: 推荐直接用PADDLE_ENFORCE_NOT_NULL

File: paddle/fluid/pybind/eager_method.cc [hunk 2]
```diff
@@ -199,17 +199,12 @@ static PyObject* tensor_clear_gradient(TensorObject* self, PyObject* args,
...
+                   paddle::platform::errors::Fatal(
+                       "Detected NULL grad"
```
Comment (chenwhql) [inline paddle/fluid/pybind/eager_method.cc:205]: Detected NULL grad后面需要标点符号断下句吗？报错结尾建议用句点结束

**Review by chenwhql — APPROVED** [review]
报错误检，细节后续可以完善下

File: paddle/fluid/pybind/eager_method.cc (hunk 1 above)
Comment (jim19930609) [reply paddle/fluid/pybind/eager_method.cc:203]: 多谢！

File: paddle/fluid/eager/api/utils/hook_utils.cc [hunk 3]
```diff
@@ -52,9 +52,15 @@ void RegisterReduceHookForTensor(const paddle::experimental::Tensor& tensor,
   }
 }
 
-void RetainGradForTensor(const paddle::experimental::Tensor& tensor) {
-  // TODO(jiabin): Support More Tensor type here
+static void RetainGradForRegularNode(
```
Comment (JiabinYang) [inline paddle/fluid/eager/api/utils/hook_utils.cc:55]: merge this with `RegisterGradientHookForTensor`

File: paddle/fluid/pybind/eager_method.cc (hunk 2 above)
Comment (jim19930609) [reply paddle/fluid/pybind/eager_method.cc:205]: Will do

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request merged**